docs/source/tutorial/Introduction.lhs

```diff
@@ -0,0 +1,65 @@
+# Defining GraphQL type APIs
+
+First some imports:
+
+``` haskell
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE OverloadedStrings #-}
+module Introduction where
+
+import Protolude
+
+import GraphQL.TypedSchema (Object, Field, Argument, (:>))
+import GraphQL.TypedApi (HandlerType, (:<>)(..))
+```
+
+The core idea for this library is that we define a composite type that
+specifies the whole API, and then implement a matching handler.
+
+The main GraphQL entities we care about are Objects and Fields. Each
+Field can have arguments.
+
+``` haskell
+type HelloWorld = Object "HelloWorld" '[]
+  '[ Argument "greeting" Text :> Field "me" Text
+   ]
+```
+
+The example above is equivalent to the following GraphQL type:
+
+```
+type HelloWorld {
+  me(greeting: String!): String!
+}
+```
+
+And if we had a code to handle that type (more later) we could query it like this:
+
+```
+{ me(greeting: "hello") }
+```
+
+## The handler
+
+We defined a corresponding handler via the `HandlerType m a` which takes
+the monad to run in (`IO` in this case) and the actual API definition
+(`HelloWorld`).
+
+```haskell
+handler :: HandlerType IO HelloWorld
+handler = pure $ (\greeting -> pure (greeting <> " to me")) :<> ()
```

Comment:
Sadly no, because I added a MonadIO constraint everywhere. It might be possible to remove that constraint, but I'm a bit unclear ATM what guarantees we need from the underlying modad.